File: backend/app/models/market.py
from sqlalchemy import Column, String, DateTime, Float
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class MarketKline(Base):
    __tablename__ = "market_klines"

    time = Column(DateTime(timezone=True), primary_key=True)
    symbol = Column(String, primary_key=True)
    interval = Column(String, primary_key=True) # 1m, 1h, 1d
    open = Column(Float)
    high = Column(Float)
    low = Column(Float)
    close = Column(Float)
    volume = Column(Float)
    source = Column(String) # binance, okx

    # --- Technical Indicators ---
    # Trend
    sma_7 = Column(Float, nullable=True)
    sma_25 = Column(Float, nullable=True)
    ema_7 = Column(Float, nullable=True)
    ema_25 = Column(Float, nullable=True)
    
    # Momentum
    rsi_14 = Column(Float, nullable=True)
    macd = Column(Float, nullable=True)
    macd_signal = Column(Float, nullable=True)
    macd_hist = Column(Float, nullable=True)
    
    # Volatility
    bb_upper = Column(Float, nullable=True)
    bb_middle = Column(Float, nullable=True)
    bb_lower = Column(Float, nullable=True)
    atr_14 = Column(Float, nullable=True)

File: frontend/src/lib/api/market.ts
import { API_BASE_URL } from "./base";

export interface KlineData {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  rsi?: number;
  macd?: number;
  macd_signal?: number;
  macd_hist?: number;
  ma20?: number;
  ma50?: number;
  sma_7?: number;
  sma_25?: number;
  ema_7?: number;
  ema_25?: number;
  bb_upper?: number;
  bb_middle?: number;
  bb_lower?: number;
  atr_14?: number;
  rsi_calc_time?: number | null;
  macd_calc_time?: number | null;
  macd_signal_calc_time?: number | null;
  macd_hist_calc_time?: number | null;
  ma20_calc_time?: number | null;
  ma50_calc_time?: number | null;
  sma_7_calc_time?: number | null;
  sma_25_calc_time?: number | null;
  ema_7_calc_time?: number | null;
  ema_25_calc_time?: number | null;
  bb_upper_calc_time?: number | null;
  bb_middle_calc_time?: number | null;
  bb_lower_calc_time?: number | null;
  atr_14_calc_time?: number | null;
}

export const MarketAPI = {
  getKline: async (symbol: string, interval: string = '1m', limit: number = 100): Promise<KlineData[]> => {
    try {
      // Use encodeURIComponent for symbol (e.g., BTC/USDT -> BTC%2FUSDT)
      const res = await fetch(`${API_BASE_URL}/api/v1/market/kline?symbol=${encodeURIComponent(symbol)}&interval=${interval}&limit=${limit}`);
      if (!res.ok) {
        console.error('API Error:', res.statusText);
        return [];
      }
      return await res.json();
    } catch (error) {
      console.error('Fetch Error:', error);
      return [];
    }
  }
};
